### ISSUE TITLE
Avoid code deduplication with attrs

### ISSUE THREAD

**Issue by guyzmo** [unfoldingWord-dev/python-gogs-client#10]
The code relies a lot on a common super-verbose pattern:

```python
class GogsFoobar:
    def __init__(self, param):
        self._param = param

    @staticmethod
    def from_json(self, parsed_json):
        param = parsed_json.get(param, 42)
        return GogsFoobar(param=param)

    @property
    def param(self):
        return self._param
```

you end up writing 9 times `param`, for each attribute you add. I love python, and things being explicit better than implicit, but we're light years away from DRY here. This is being…

![](http://bartsblackboard.com/files/2010/01/the_simpsons.10x23.30_minutes_over_tokyo.jpg)

I suggest instead to use some tricks like, the [`attrs`](https://attrs.readthedocs.io/en/stable/) module:

```python
import attr

@attr.s(frozen=True)
class GogsFoobar:
    foobar = attr.ib(default=42)

    @staticmethod
    def from_json(self, parsed_json):
        param = parsed_json.get(param, 42)
        return GogsFoobar(param=param)
```

it's now down to 4.

But we can make it totally DRY, the following way:

```python
import attr

@attr.s
class GogsEntity:
    json = attr.ib()
    @classmethod
    def from_json(cls, parsed_json):
        # with introspection, get arguments of the constructor
        params = attr.fields(cls)
        args = []
        kwargs = {}
        for param in params:
            # if not a keyword argument
            if param.name == attr._make._Nothing:
                args.append(json_get(parsed_json, param.name))
            # if it's a keyword argument
            else:
                kwargs[param.name] = parsed_json.get(param.name, None)
        del kwargs['json']
        instance = cls(*args, json=parsed_json, **kwargs)
        return instance

@attr.s(frozen=True)
class GogsFoobar(GogsEntity):
    foo = attr.ib()
    foobar = attr.ib(default=42)

x = GogsFoobar.from_json(dict(foo=1, foobar=2))

print(x.foo, x.foobar, x.json)
```

and now you only write the attribute once. But I agree, it's using some creepy dirty internal automagic from python. At the cost of making it uglier in one place, you make it DRY everywhere else, with the same strong type checkings you enforce currently.

**Comment by guyzmo** [comment]
(if you checked you email first, note that I updated the issue to make the code better and working ☺)

**Comment by ethantkoenig** [comment]
Resolved by d3db08e5850471f79f83c22b240823fb8191779c
